File: mobile/src/main/kotlin/de/scout/fireplace/ui/DynamicHeightImageView.java
package de.scout.fireplace.ui;

import android.content.Context;
import android.content.res.TypedArray;
import android.support.v7.widget.AppCompatImageView;
import android.util.AttributeSet;
import de.scout.fireplace.R;

public class DynamicHeightImageView extends AppCompatImageView {

  private double heightRatio;

  public DynamicHeightImageView(Context context) {
    super(context);
    init(context, null);
  }

  public DynamicHeightImageView(Context context, AttributeSet attrs) {
    super(context, attrs);
    init(context, attrs);
  }

  public DynamicHeightImageView(Context context, AttributeSet attrs, int defStyle) {
    super(context, attrs, defStyle);
    init(context, attrs);
  }

  @Override
  protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
    if (heightRatio > 0.0) {
      int width = MeasureSpec.getSize(widthMeasureSpec);
      int height = (int) (width * heightRatio);
      setMeasuredDimension(width, height);
    } else {
      super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }
  }

  private void init(Context context, AttributeSet attrs) {
    if (!isInEditMode()) {
      TypedArray attributeArray = context.obtainStyledAttributes(attrs, R.styleable.DynamicHeightImageView);

      try {
        float heightRatio = attributeArray.getFloat(R.styleable.DynamicHeightImageView_heightRatio, 1.0F);
        this.heightRatio = heightRatio;
      } finally {
        attributeArray.recycle();
      }
    }
  }

  public void setHeightRatio(double ratio) {
    if (ratio != heightRatio) {
      heightRatio = ratio;
      requestLayout();
    }
  }

  public double getHeightRatio() {
    return heightRatio;
  }
}

File: feature/src/main/kotlin/de/scout/fireplace/feature/ui/DynamicHeightImageView.kt
package de.scout.fireplace.feature.ui

import android.content.Context
import android.support.v7.widget.AppCompatImageView
import android.util.AttributeSet
import android.view.View
import de.scout.fireplace.feature.R

internal class DynamicHeightImageView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyle: Int = 0) : AppCompatImageView(context, attrs, defStyle) {

  private var heightRatio: Double = 0.toDouble()

  init {
    if (!isInEditMode) {
      val attributeArray = context.obtainStyledAttributes(attrs, R.styleable.DynamicHeightImageView)

      try {
        val heightRatio = attributeArray.getFloat(R.styleable.DynamicHeightImageView_heightRatio, 1.0f)
        this.heightRatio = heightRatio.toDouble()
      } finally {
        attributeArray.recycle()
      }
    }
  }

  override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
    if (heightRatio > 0.0) {
      val width = View.MeasureSpec.getSize(widthMeasureSpec)
      val height = (width * heightRatio).toInt()
      setMeasuredDimension(width, height)
    } else {
      super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    }
  }
}
